File: 160706/number/ruby/seven.rb
# 7. 实现一个方法，传入数字n，顺序打印1到n中所有非质数，并做质因数分解

num = 9
# 将质数和非质数存在数组中
def prime_analyze(num)
  prime_ary = []
  not_prime_ary = []
  for i in 1..num
    flag = false
    for j in 1..i
      if j != 1 && j != i
        if i % j ==0
          flag = true
          break
        end
      end
    end
    if flag == true
      not_prime_ary.push(i)
    end
    if flag == false
      prime_ary.push(i)
    end
  end
  prime_ary.delete(1)
  not_prime_ary.each do |x|
    print "开始对#{x}进行质因数分解：#{x}="
    # 调用质因数分解的方法
    prime_factorizations(x, prime_ary)
    p "结束"
  end
end

# 质因数分解的方法
def prime_factorizations(not_prime, prime_ary)
  flag = true
  for i in 0..prime_ary.length - 1
    y = prime_ary[i]
    break if not_prime < y
    if not_prime % y == 0
      if flag == true
        print y
        flag = false
      else
        print "*#{y}"
      end
      not_prime = not_prime / y
      redo
    end
  end
end

# 运行分解质数和非质数的方法
prime_analyze(num)

File: 160706/number/java/seven.java
// 7. 实现一个方法，传入数字n，顺序打印1到n中所有非质数，并做质因数分解
import java.util.Arrays;
import java.util.ArrayList;

public class seven{
  public static void main(String[] args) {
    int num = 9;
    // 调用分解质数和非质数的方法
    primeFactorizations(num);
  }
  
  /*
  * 将质数和非质数分解出来存在数组中
  */
  public static void primeFactorizations(int num){
    ArrayList<Integer> notPrimeAry = new ArrayList<Integer>();
    ArrayList<Integer> primeAry = new ArrayList<Integer>();
    for (int i = 1; i <= num; i ++) {
      boolean flag = false;
      for (int j = 1; j <= i; j ++) {
        if (j != 1 && j != i) {
          if (i % j == 0) {
            flag = true;
            break;
          }
        }
      }
      if (flag == true) {
        notPrimeAry.add(i);
      }
      if (flag == false && i != 1) {
        primeAry.add(i);
      }
    }
    // 调用分解质因数的方法
    primeAnalyze(notPrimeAry, primeAry);
  }

  /*
  *  进行质因数分解
  */
  public static void primeAnalyze(ArrayList<Integer> notPrimeAry, ArrayList<Integer> primeAry){
    for (int k = 0; k < notPrimeAry.size(); k ++) {
      int qualityNum = notPrimeAry.get(k);
      boolean tag = true;
      System.out.print("对" + qualityNum + "进行质因数分解：" + qualityNum + " = " );
      for (int p = 0; p < primeAry.size(); p ++) {
        int n = primeAry.get(p);
        if (qualityNum < n)
          break;
        if (qualityNum % n == 0) {
          if (tag == true) {
            System.out.print(n);
            tag = false;
          }else{
            System.out.print("*" + n);
          }
          p--;
          qualityNum = qualityNum / n;
        }
      }
      System.out.println(" 结束");
    }
  }
}
